src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp

```diff
@@ -0,0 +1,185 @@
+/**
+ * @file bilinear_interpolation_impl.hpp
+ * @author Kris Singh and Shikhar Jaiswal
+ *
+ * Implementation of the bilinear interpolation function as an individual layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_BILINEAR_INTERPOLATION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BILINEAR_INTERPOLATION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "bilinear_interpolation.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+template<typename InputDataType, typename OutputDataType>
+BilinearInterpolation<InputDataType, OutputDataType>::
+BilinearInterpolation():
+    inRowSize(0),
+    inColSize(0),
+    outRowSize(0),
+    outColSize(0),
+    depth(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+BilinearInterpolation<InputDataType, OutputDataType>::
+BilinearInterpolation(
+    const size_t inRowSize,
+    const size_t inColSize,
+    const size_t outRowSize,
+    const size_t outColSize,
+    const size_t depth):
+    inRowSize(inRowSize),
+    inColSize(inColSize),
+    outRowSize(outRowSize),
+    outColSize(outColSize),
+    depth(depth)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BilinearInterpolation<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, 1);
+  else
+  {
+    assert(output.n_rows == outRowSize * outColSize * depth);
+    assert(output.n_cols == 1);
+  }
+
+  assert(inRowSize >= 2);
+  assert(inColSize >= 2);
+
+  arma::cube inputAsCube(input.memptr(), inRowSize, inColSize, depth);
```

Comment:
If `false` isn't specified as the fifth argument, this will by default copy all of the data in the cube.  Are you sure this is what you want to do here?

Reply:
This is actually using `cube(const ptr_aux_mem, n_rows, n_cols, n_slices)` instead of `cube(ptr_aux_mem, n_rows, n_cols, n_slices, copy_aux_mem = true, strict = false)` as `input` is of `const` type.

Reply:
Ah right, `input` is const, now it makes sense.  Thanks for the clarification.